Migrate to new repo

reginleif888 opened the pull request.

reginleif888 committed:
Move source code to new repo

reginleif888 committed:
Update README.md

reginleif888 committed:
Update mops.toml

reginleif888 committed:
Some doc comments updates

reginleif888 committed:
Edit CI

reginleif888 committed:
Improve doc comments

reginleif888 committed:
Add example code

reginleif888 committed:
Rename Mapping.mo to util.mo

reginleif888 committed:
Edit example

timohanke commented on src/AccountManager.mo:
```diff
@@ -0,0 +1,597 @@
+      return ?inc;
+    };
+
+    func processDepositTransfer(account : ICRC1.Account, amount : Nat) : async* {
```
The function name is close to the `process_deposit` above. 

Maybe `process_allowance` is better to distinguish it?

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+    withdrawal_fee : Nat;
+  };
+
+  type NotifyResult = {
+    #Ok : {
+      deposit_inc : Nat;
+      credit_inc : Nat;
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #NotAvailable : Text;
+    };
+  };
+
+  type WithdrawResult = {
+    #Ok : {
+      txid : Nat;
+      amount : Nat;
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #InsufficientCredit;
+      #AmountBelowMinimum;
+    };
+  };
```
We should export these type from TokenHandler so that the example code does not have to define them.

For consistent naming better to call these NotifyRespose instead of NotifyResult.

reginleif888 replied:
Naming is taken from the standard: https://github.com/research-ag/icrcX/blob/main/icrcX.md#notification
Regarding types reuse, we should define them locally, because these types are not related to the token handler.

timohanke replied:
You are right. The types are related to icrcX, not to TokenHandler. So we could make a file icrcX.mo in the future to import these files.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+    assets := Vec.map<StableAssetInfo, AssetInfo>(
+      assetsData,
+      func(x) {
+        let r = (actor (Principal.toText(x.ledgerPrincipal)) : ICRC1.ICRC1Ledger)
+        |> {
+          balance_of = _.icrc1_balance_of;
+          fee = _.icrc1_fee;
+          transfer = _.icrc1_transfer;
+          transfer_from = _.icrc2_transfer_from;
+        }
+        |> {
+          ledgerPrincipal = x.ledgerPrincipal;
+          handler = TokenHandler.TokenHandler(_, Principal.fromActor(self), JOURNAL_SIZE, 0, true);
+        };
```
It looks like we need some convenience functions to reduce this boiler plate code for the developer.

For example, we could provide a convenience function in ICRC1.mo which takes either a principal or the actor and produces the type LedgerAPI which can then be passed as the first argument to the TokenHandler constructor.

Or we need a second constructor in lib.mo which wraps around TokenHandler and which takes a principal instead of type LedgerAPI as first argument and does the rest internally.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+    #Err : { #NotAvailable : Text };
+  } {
+    assertInitialized();
+    (
+      switch (getAssetInfo(token)) {
+        case (?aid) aid;
+        case (_) return #Err(#NotAvailable("Unknown token"));
+      }
+    )
```
If the null case leads to a `return` then a `let-else` statement is often nicer. Can't use the pipe then but still. 

The icrcX spec says to throw an error (i.e. `throw Error.reject("..")`) in case the token isn't known.

For example
```
let ?assetInfo = getAssetInfo(token) else throw Error.reject("Unknown token");
```

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+    #Ok : Nat;
+    #Err : { #NotAvailable : Text };
+  } {
+    assertInitialized();
```
The icrcX spec says to throw an error (i.e. `throw Error.reject("..")`) in case the token isn't known. It does not return a Result type, i.e. no #Ok/#Err variant.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+
+  public shared query ({ caller }) func icrcX_credit(token : Principal) : async Int {
+    assertInitialized();
+    switch (getAssetInfo(token)) {
+      case (?info) info.handler.userCredit(caller);
+      case (_) 0;
+    };
```
Can use `let-else` instead.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+
+  public shared ({ caller }) func icrcX_notify(args : { token : Principal }) : async NotifyResult {
+    assertInitialized();
+    let assetInfo = switch (getAssetInfo(args.token)) {
+      case (?aid) aid;
+      case (_) return #Err(#NotAvailable("Unknown token"));
+    };
```
Can use let-else instead

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+      case (_) return #Err(#NotAvailable("Unknown token"));
+    };
+    let result = try {
+      ignore await* assetInfo.handler.fetchFee();
```
This line can be deleted. `handler.notify()` handles fee internally.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+        #Ok({ deposit_inc = delta; credit_inc = delta });
+      };
+      case (null) {
+        #Ok({
+          deposit_inc = 0;
+          credit_inc = 0;
+        });
+      };
```
icrcX spec says to return `#Err(#NotAvailable)` in this case.

timohanke replied:
icrcX spec says to return #Err(#NotAvailable) in this case.

timohanke commented on example/main.mo:
```diff
@@ -0,0 +1,255 @@
+    };
+    switch (result) {
+      case (?(delta, usableBalance)) {
+        await* assetInfo.handler.trigger(1);
```
We can remove this line. TokenHandler was configured with auto-trigger = true.

reginleif888 committed:
Token handler edits (#3)

timohanke reviewed (commented):
Looks good. Just waiting for additional tasks to be merged in here before we merge this into main.

timohanke committed:
set trigger(n) to n=10 in example (#5)

In the example we set the timer to 60 seconds. The trigger runs n
sequential consolidations. Each consolidation as a cross-subnet
call takes about 4-5 seconds normally. So we should be able to
complete 10 in a minute. If there is a large backlog then it is
better to process as much as possible of it.

reginleif888 committed:
New updates (#6)

* TokenHandler named constructor parameters

* Turn all error variants into records in example

* Validate withdrawal subaccount

* Use icrc84 name

* Update response from notify/deposit in example

reginleif888 committed:
Unique spender_subaccount for the user

reginleif888 committed:
Fix deposit via an allowance

timohanke reviewed (commented):
For the last commit we could use a test of`handler.depositFromAllowance(p, account, amount)` where the caller principal `p` is different from the principal in `account`. Just to test that internally the right principal gets credited.

reginleif888 committed:
Fix credit registry share/unshare

reginleif888 committed:
Improvements (#9)

* depositFromAllowance() improvements
* Update example according to icrc-84 changes
* Fix LogEvent types
* Add test for depositFromAllowance() when caller principal != account owner
* Add notify_tag_commit_hook.yml
* Refactor BadFee retries
* Refactor error types
* Wrap a class around ledger api and abstract away the async calls into separate file

timohanke committed:
Make example comply with icrc-84.did and check that in CI (#11)

reginleif888 committed:
notify_on_tag_creation.yml

timohanke commented on .github/workflows/notify_on_tag_creation.yml:
```diff
@@ -0,0 +1,40 @@
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Set up environment variables
+        run: |
+          echo "NOTION_DATABASE_ID=e3aca840563b4cc8b273e65da748e1dd" >> $GITHUB_ENV
+          echo "WEBHOOK_URL=https://hook.eu2.make.com/sg1xhhlvnp4314llgde92csqgjrpm42e" >> $GITHUB_ENV
```
Can we remove these lines now because we have this at the top?
```
env:
  NOTION_DATABASE_ID: e3aca840563b4cc8b273e65da748e1dd
  WEBHOOK_URL: https://hook.eu2.make.com/sg1xhhlvnp4314llgde92csqgjrpm42e
```

timohanke commented on .github/workflows/notify_on_tag_creation.yml:
```diff
@@ -0,0 +1,40 @@
+on:
+  create:
+    tags:
+      - "rc*"
```
Have you tested that this tag pattern works as expected?

timohanke reviewed (commented):
Left one request about notify hook

reginleif888 committed:
Update notify_on_tag_creation.yml

reginleif888 committed:
Prerelease improvements (#13)

* Test journal events

* Withdrawal improvements

* Remove minima

* Introduce surcharge

* Update doc comments

* Introduce new integrity assertion

* Take fee difference into account in consolidation

* Rework deposit via allowance

* Introduce the credit registry cleanup

* Rework notifications tests

* Test surcharge to notifications tests

* Rework consolidation tests

* Rework deposit via allowance tests

* Add test for multiple consolidations trigger

* Update allowance tests

* Rework withdrawal tests

* Fix allowance tests

* Move all other tests to a separate file

* Remove unnecessary files

* Test surcharge and fees

* Fix doc comment for credited

* Update example

* Rework README.md

* Update doc comments

* Update icrc84.did (#14)

example/mops.toml now uses icrc84#0.0.1
CI automatically uses icrc84.did from example/mops.toml

* use icrc-84 repo with empty src directory (#15)

This satisfies `moc` when looking at the icrc-84 "module" installed by mops. Without the `src` directory we will see warnings in VSCode.

Also fix tag_action which accidentally triggered

* Fix clean() return type

* Mention multiple token support in README

* Fix typo

* Remove #LedgerBadFee error variant

* Improve allowance

* Remove clean() functionality

---------

Co-authored-by: Timo Hanke <t@timohanke.org>

AStepanov25 committed:
Rewrite tests using AsyncVariableTester. (#17)

* Rewrite tests using AsyncVariableTester.

* Remove old mock_ledger file.

* Remove duplicate definitions.

* Rewrite createHandlerV2

* Fix tests.

* Remove redundant function.

AStepanov25 committed:
Fix tests with stage_unlocked.

timohanke commented:
Abandoning this PR in favor of #20 . Work will continue on #20
